core-foundation/src/filedescriptor.rs

```diff
@@ -0,0 +1,71 @@
+pub use core_foundation_sys::filedescriptor::*;
+
+use core_foundation_sys::base::{Boolean, CFIndex, CFRelease};
+use core_foundation_sys::base::{kCFAllocatorDefault, CFOptionFlags};
+
+use base::{TCFType};
+
+use std::os::unix::io::{AsRawFd, RawFd};
+
+pub struct CFFileDescriptor(CFFileDescriptorRef);
+
+impl Drop for CFFileDescriptor {
+    fn drop(&mut self) {
+        unsafe {
+            CFRelease(self.as_CFTypeRef())
+        }
+    }
+}
+
+impl_TCFType!(CFFileDescriptor, CFFileDescriptorRef, CFFileDescriptorGetTypeID);
+
+impl CFFileDescriptor {
+    pub unsafe fn new(fd: RawFd,
+                      closeOnInvalidate: bool,
+                      callout: CFFileDescriptorCallBack,
+                      context: *const CFFileDescriptorContext) -> CFFileDescriptor {
+        let fd_ref = CFFileDescriptorCreate(kCFAllocatorDefault,
```

Comment:
This can return null if there's a problem, according to [the docs](https://developer.apple.com/documentation/corefoundation/1477591-cffiledescriptorcreate?language=objc).